File: duna-core/src/main/java/io/duna/core/proxy/RemoteServiceCallInterceptor.java
/*
 * Copyright (c) 2016 Duna Open Source Project
 * Ministério do Planejamento, Desenvolvimento de Gestão
 * República Federativa do Brasil
 *
 * This file is part of the Duna Project.
 */
package io.duna.core.proxy;

import io.duna.core.DunaException;
import io.duna.core.io.BufferInputStream;
import io.duna.core.io.BufferOutputStream;
import io.duna.core.service.ServiceException;
import io.duna.core.util.Services;

import co.paralleluniverse.fibers.SuspendExecution;
import com.fasterxml.jackson.core.JsonEncoding;
import com.fasterxml.jackson.core.JsonGenerator;
import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.google.common.primitives.Primitives;
import io.vertx.core.Vertx;
import io.vertx.core.VertxException;
import io.vertx.core.buffer.Buffer;
import io.vertx.core.eventbus.Message;
import net.bytebuddy.implementation.bind.annotation.*;

import java.io.IOException;
import java.lang.reflect.Method;
import java.util.HashMap;
import java.util.Map;
import java.util.logging.Logger;

import static io.vertx.ext.sync.Sync.awaitResult;

public class RemoteServiceCallInterceptor {

    private final Map<String, String> addressCache = new HashMap<>();

    @RuntimeType
    public Object intercept(@FieldValue("logger") Logger logger,
                            @FieldValue("objectMapper") ObjectMapper objectMapper,
                            @Origin Method method,
                            @This ServiceProxy proxyInstance,
                            @AllArguments Object ... args)
        throws SuspendExecution, InterruptedException {

        logger.fine(() -> "Intercepting call to " + method.toString());

        if (Vertx.currentContext() == null) {
            logger.severe(() -> "Proxy called outside vert.x context");
            throw new IllegalStateException("Proxy called outside vert.x context.");
        }

        String serviceAddress = addressCache.computeIfAbsent(proxyInstance.getQualifier() + method.toString(),
            k -> proxyInstance.getQualifier().isEmpty()
                ? Services.getInternalServiceAddress(method)
                : proxyInstance.getQualifier() + "@" + Services.getInternalServiceAddress(method));

        Vertx vertx = Vertx.currentContext().owner();
        BufferOutputStream outputStream = new BufferOutputStream(Buffer.buffer());

        try {
            JsonGenerator generator = objectMapper.getFactory()
                .createGenerator(outputStream, JsonEncoding.UTF8);

            generator.writeStartArray(args.length);
            for (Object arg : args) generator.writeObject(arg);
            generator.writeEndArray();

            generator.flush();
            generator.close();

            logger.info(() -> "Sending request to " + serviceAddress);

            Class<?> wrapperReturnType = Primitives.wrap(method.getReturnType());

            if (Void.class.isAssignableFrom(wrapperReturnType)) {
                vertx.eventBus().send(serviceAddress, outputStream.getBuffer());
                return null;
            } else {
                try {
                    Message<Buffer> response = awaitResult(
                        h -> vertx.eventBus().send(serviceAddress, outputStream.getBuffer(), h));

                    if (response.body() != null && response.body().length() > 0) {
                        BufferInputStream inputStream = new BufferInputStream(response.body());
                        JsonParser parser = objectMapper.getFactory().createParser(inputStream);

                        final Object result = parser.readValueAs(method.getReturnType());

                        parser.close();
                        outputStream.close();
                        inputStream.close();
                        return result;
                    } else {
                        throw new DunaException("The request resulted in an empty answer.");
                    }
                } catch (VertxException ex) {
                    if (ex.getCause() instanceof ServiceException) {
                        throw (ServiceException) ex.getCause();
                    } else if (ex.getCause() instanceof DunaException) {
                        throw (DunaException) ex.getCause();
                    } else {
                        throw new DunaException(ex.getCause());
                    }
                }
            }
        } catch (IOException ex) {
            throw new DunaException(ex);
        }
    }
}

File: duna-core/src/main/kotlin/io/duna/core/proxy/RemoteServiceCallInterceptor.kt
/*
 * Copyright (c) 2016 Duna Open Source Project
 * Ministério do Planejamento, Desenvolvimento de Gestão
 * República Federativa do Brasil
 *
 * This file is part of the Duna Project.
 */
package io.duna.core.proxy

import co.paralleluniverse.fibers.SuspendExecution
import com.fasterxml.jackson.databind.ObjectMapper
import io.duna.core.DunaException
import io.duna.io.BufferInputStream
import io.duna.io.BufferOutputStream
import io.duna.util.Services
import io.vertx.core.Vertx
import io.vertx.core.buffer.Buffer
import io.vertx.core.eventbus.DeliveryOptions
import io.vertx.core.eventbus.Message
import io.vertx.ext.sync.Sync
import net.bytebuddy.implementation.bind.annotation.AllArguments
import net.bytebuddy.implementation.bind.annotation.FieldValue
import net.bytebuddy.implementation.bind.annotation.Origin
import net.bytebuddy.implementation.bind.annotation.RuntimeType
import java.lang.reflect.Method
import java.nio.charset.Charset
import java.util.concurrent.ConcurrentHashMap
import java.util.logging.Logger

/**
 * Interceptor used to forward service calls to remote instances.
 *
 * @author Carlos Eduardo Melo <[ceduardo.melo@gmail.com]>
 */
internal object RemoteServiceCallInterceptor {

  val addressCache = ConcurrentHashMap<Method, String>()

  @RuntimeType
  @Throws(SuspendExecution::class, InterruptedException::class)
  fun intercept(@FieldValue("logger") logger: Logger,
                @FieldValue("objectMapper") objectMapper: ObjectMapper,
                @Origin(cache = true) method: Method,
                @AllArguments vararg arguments: Any?): Any? {

    logger.fine { "Intercepting method $method" }

    val vertx = Vertx.currentContext()?.owner() ?:
      throw DunaException("There isn't a vert.x context available.")

    val outBuffer = BufferOutputStream(Buffer.buffer(1024))
    val generator = objectMapper.factory.createGenerator(outBuffer)

    generator.writeStartObject()
    arguments.forEachIndexed { i, any -> generator.writeObjectField("$i", any) }
    generator.writeEndObject()

    generator.flush()
    generator.close()

    val serviceAddress = addressCache[method] ?:
      Services.getInternalServiceAddress(method)

    addressCache[method] = serviceAddress

    val deliveryOptions = DeliveryOptions()
    deliveryOptions.addHeader("action", method.name)
    deliveryOptions.addHeader("paramTypes",
      method.parameterTypes
        .map { it.toGenericString() }
        .joinToString(",")
    )

    // TODO Add support for request filters

    logger.info { "Sending request to $serviceAddress" }
    logger.finer { "Request: ${outBuffer.buffer.toString(Charset.defaultCharset())}"}

    val response = Sync.awaitResult<Message<Buffer>> {
      vertx.eventBus().send(serviceAddress, outBuffer.buffer, it)
    }

    logger.fine { "Response received from $serviceAddress" }
    logger.finer { "Response contents: ${response.body().toString(Charset.defaultCharset())}" }

    if (response.body() != null && response.body().length() > 0) {
      val inBuffer = BufferInputStream(response.body())
      val parser = objectMapper.factory.createParser(inBuffer)
      val result = parser.readValueAs(method.returnType)

      parser.close()
      return result
    }

    if (method.returnType.isAssignableFrom(Void::class.java) ||
      method.returnType.isAssignableFrom(Unit::class.java))
      return Unit
    else
      return null
  }
}
